File: go/26.remove-duplicates-from-sorted-array.go
/*
 * @lc app=leetcode id=26 lang=golang
 *
 * [26] Remove Duplicates from Sorted Array
 */
func removeDuplicates(nums []int) int {
    return naive(nums)
}

func naive(nums []int) int {
    length := len(nums)
    start := 0
    count := 0
    for i := 1; i < length; i++ {
        if nums[start] == nums[i] {
            count++
        } else {
            if count != 0 {
                shiftLeft(nums, length, start+1, count)
                length -= count
                i -= count
            }
            start = i
            count = 0
        }
    }
    if count != 0 {
        shiftLeft(nums, length, start+1, count)
        length -= count
    }
    return length
}

func shiftLeft(nums []int, length int, start int, count int) {
    for i := start; i+count < length; i++ {
        nums[i] = nums[i+count]
    }
}

File: csharp/26.remove-duplicates-from-sorted-array.cs
/*
 * @lc app=leetcode id=26 lang=csharp
 *
 * [26] Remove Duplicates from Sorted Array
 */

// @lc code=start
public class Solution 
{
    public int RemoveDuplicates(int[] nums) 
    {
        if (nums.Length <= 1)
        {
            return nums.Length;
        }

        int newIndex = 1;

        for (var i = 1; i < nums.Length; i++)
        {
            if (nums[i-1] != nums[i])
            {
                nums[newIndex++] = nums[i];
            }
        }

        return newIndex;
    }
}
// @lc code=end

